### PR TITLE
modify yardopts parsing to ignore hash and c-style comments

### PR DESCRIPTION
**Pull request by eprothro**
# Description

Adds support for hash and c-style comments in .yardopts files.

We want this to be able to reduces developer confusion for those new to YARD. Devs can tend to see a .yardopts file, get confused, then just shrug and walk away. 

Being able to comment the options can pull them in more easily, and allow for documenting options that may not be use currently but are of interest.

#### Example
```
# YARD options, see `yardoc -h`
# for valid options and descriptions

# parse inline documentation with
# rdoc markup formatting
--markup "rdoc"

# treat code mixed in via include
# and extend as first class citizens
# of the module without additional directives
--embed-mixins

# clear the yardoc database
# when requesting documentation
# --no-cache

# Additional files to add to documentation tree
-
CHANGELOG.md
CONTRIBUTING.md
LICENSE
```

# Completed Tasks

* [y] I have read the [Contributing Guide][contrib].
* [y] The pull request is complete (implemented / written).
* [y] Git commits have been cleaned up (squash WIP / revert commits).
* [y] I wrote tests and ran `bundle exec rake` locally (if code is attached to PR).

[contrib]: https://github.com/lsegal/yard/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: lib/yard/cli/yardopts_command.rb [hunk 1]
```diff
@@ -65,11 +65,14 @@ def yardopts_options(opts)
...
 
-      # Parses the .yardopts file for default yard options
+      # Parses the .yardopts file for default yard options,
+      # ignoring comment lines beginning with // or #
       # @return [Array<String>] an array of options parsed from .yardopts
       def yardopts(file = options_file)
         return [] unless use_yardopts_file
-        File.read_binary(file).shell_split
+        contents = File.read_binary(file)
+        opts = contents.gsub(%r{(#|//).*\n?}, "")
```
Comment (lsegal) [inline lib/yard/cli/yardopts_command.rb]: We're going to need a requirement that comments start at the beginning of the line (incl. optional whitespace), otherwise a valid option like the following will not work:

```
--title "# a title!"
```

Or even:

```
lib/*/*.rb
lib//*.rb
```

Unfortunately this would be a hard requirement.

I would recommend tests for these edge cases too.

File: spec/cli/yardoc_spec.rb [hunk 2]
```diff
@@ -510,10 +510,23 @@ def foo; end
       optsdata = String.new("foo bar")
       expect(optsdata).to receive(:shell_split)
       expect(File).to receive(:read_binary).with("test").and_return(optsdata)
+      allow(optsdata).to receive(:gsub).and_return(optsdata)
       @yardoc.options_file = "test"
       @yardoc.run
     end
 
+    it "ignores .yardopts tokens that are comments" do
+      optsdata = String.new("--one-file --locale es # --locale en\n// --title not_my_title")
```
Comment (lsegal) [inline spec/cli/yardoc_spec.rb]: This test should not pass, see above. Having "#" as an argument to YARD is completely valid.

File: spec/cli/yardoc_spec.rb [hunk 3]
```diff
@@ -510,10 +510,23 @@ def foo; end
       optsdata = String.new("foo bar")
       expect(optsdata).to receive(:shell_split)
       expect(File).to receive(:read_binary).with("test").and_return(optsdata)
+      allow(optsdata).to receive(:gsub).and_return(optsdata)
```
Comment (lsegal) [inline spec/cli/yardoc_spec.rb]: Why is this mocked?

File: spec/cli/yardoc_spec.rb (hunk 2 above)
Comment (lsegal) [inline spec/cli/yardoc_spec.rb]: Another bunch of completely valid examples:

```
--default-return #read
```

```
--query "object.path.include?('#')"
```

File: spec/cli/yardoc_spec.rb (hunk 3 above)
Comment (eprothro) [inline spec/cli/yardoc_spec.rb]: The `optsdata` string is frozen in this any many other tests.  Since shell split is an extension on `String`  I can't mock there.

File: lib/yard/cli/yardopts_command.rb (hunk 1 above)
Comment (eprothro) [inline lib/yard/cli/yardopts_command.rb]: I'm completely for the constraint that this only support comments starting at the beginning of the line.

I'm happy to add more tests -- wanted to get your temperature on this first.

File: spec/cli/yardoc_spec.rb (hunk 2 above)
Comment (eprothro) [inline spec/cli/yardoc_spec.rb]: I added the "handle comments ending a line" functionality as an edge case at the last minute. Didn't think about these -- of course, you're right. Happy to not have this.

**Comment by lsegal** [comment]
> [!NOTE]
> This is a batch message, apologies if this feels impersonal.

I am closing out a few stale issues and PRs in the YARD repo and this was flagged for automatic closure because the issue may have been fixed or is likely not planned due to the issue/PR age. If this is still an issue or you believe the issue/PR is still worth a discussion, please feel free to comment or reopen a new issue. Thanks and sorry for not getting around to this the first time!
